### DISCUSSION TITLE
Another case of 403 Errors.

### DISCUSSION THREAD

**Discussion by digamesystems** [weather-gov/api#763] [General]
I've read through some of the discussions around receiving 403 errors. But I'm having trouble solving our current problem with them. 

Our app hits the API periodically to gather data for a number of locations. The API and app have been playing flawlessly together for months now but this weekend we've been shut down by 403 errors. Hitting the API from the server using curl yields: 

`$ curl -X GET "https://api.weather.gov/gridpoints/MTR/99,82/forecast" -H "accept: application/geo+json" -H "Feature-Flags: "`
<HTML><HEAD>
<TITLE>Access Denied</TITLE>
</HEAD><BODY>
<H1>Access Denied</H1>

You don't have permission to access "http&#58;&#47;&#47;api&#46;weather&#46;gov&#47;gridpoints&#47;MTR&#47;99&#44;82&#47;forecast" on this server.<P>
Reference&#32;&#35;18&#46;ca7cb17&#46;1725143120&#46;2fec2243
<P>https&#58;&#47;&#47;errors&#46;edgesuite&#46;net&#47;18&#46;ca7cb17&#46;1725143120&#46;2fec2243</P>
</BODY>
</HTML>

The EdgeSuite link just returns an error. Have we been identified as a bot? Possibly making too many calls in a period of time? Any help would be greatly appreciated.

**Comment by jah1007** [comment]
Took a look at the error string and it appears that the IP address you were cycled into was previously used by a bad actor and is triggering security controls. What we have seen is that it typically takes a varying amount of time for the IP address to no longer be seen as abusive. Our only suggestion is to try to obtain a static IP from your ISP.

**Reply by digamesystems** [reply]
Thanks, jah1007. 

I reached out to our hosting service (Akamai) and they responded to your note: 

> Each Linode _is_ assigned a static IP address, which they will keep for the duration of their existence. The only time a Linode's IP address would change is if it were manually swapped out by the user, or if the Linode is migrated to a different data center (as IPs are restricted to the data center at which they are issued from).
> 
> Looking at your Linode "qpe-assistant", I see that it was assigned its IP address xxx.xxx.xxx.xxx on the January 11, 2024 (the date of its creation) and has been using this IP since.
> 
> With that in mind, did the NWS provide any information about the nature of what was triggering those security controls or how long it can take for them to no longer see the IP as abusive? With this IP address belonging to your Linode for the past 8 months or so, I would think that would be long enough for them remove any restrictions for the IP if it had previously been used by a bad actor previous to January 11, 2024. I am in no way trying to infer that any of your behavior while using this IP has caused these security triggers to eventually block the IP, I am just trying to get an idea of NWS's process here.
> 
> If you had also been able to connect to the NWS API successfully, I am wondering how long you were able to make API calls before they placed restrictions on the IP? You may be able to inform them of the date that this IP address was assigned to you, and let them know that if any "bad acting" before this date is the reason for the block, then that person is longer using this IP address.
> 
> The tricky thing about this type of situation is that, attempting to switch out your Linodes IP address may not be successful, as we have no way of knowing if that new IP will also be blocked for NWS's system. Since we have no idea of their criteria for blocking an IP, we are unable to assign a new IP from our end in cases like these.

----
Our application hits the NWS API about every half hour and looks for updates to the forecasts for our customers' sites. We are working with a beta community of about 40 users for early feedback, so I would hope the amount of traffic we are generating is relatively modest. Since the problem just came up, and we've had the IP address for some time, I wonder if we've been classified as a 'abusive'...? 

What's the right support channel for us within NWS to help run this down? "nco.ops@noaa.gov" <nco.ops@noaa.gov>? 

Thanks again,
John

**Reply by jah1007** [reply]
We also use Akamai so I will forward this on to our client rep and see if they can figure out something.

**Reply by jaredwsmith** [reply]
I, too, am on Akamai (Linode) and have found my server blocked. I've had this IP address for years. I'm hosting in the Atlanta data center if this narrows it down any.

**Reply by digamesystems** [reply]
Interesting. I've got a ticket started with the weather.gov team and linode. Anyone else with this issue might want to reach out to nco.ops@nooa.gov as well as linode and help us all run this down.

**Reply by digamesystems** [reply]
My linode server is running out of Fremont, Ca

**Reply by cirrustelecom** [reply]
I also have been blocked, been using the same IP for almost 12 years now out of Fremont Linode.  I contacted nco.ops, no response yet.  We also have an MOU with DHS and we have been blocked from IPAWS as well.

**Reply by dwhitemv25** [reply]
Can you try accessing the API over IPv6? If Akamai/Linode has rolled out CGNAT then this may become a recurring problem. With IPv6 you should be able to get a unique address.

**Reply by cirrustelecom** [reply]
I have a static ipv6 from linode.  Using ipv6 to hit the API.  Tried with ipv4 and got 403.

**Reply by VioletPixel** [reply]
Getting the same thing, also on Linode.  It's been working fine for years on the same IP.  Sent an email, we'll see what happens.

**Reply by digamesystems** [reply]
Status update: I've got tickets open with both the NOAA team and Linode. -- Linode says there isn't much they can do about NOAAs blocking policies. I'll grant that, but it might be good to get some kind of mapping of the problem in terms of affected IPs / Domains. My server lives at: 104.237.152.51 (QPEAssistant.com) I sent the NOAA support team a link to this thread so hopefully, they'll be checking in on it sometime today.

**Reply by hanksims** [reply]
All the same problems here, again with a Fremont Linode.

**Reply by digamesystems** [reply]
If anyone else is putting tickets into NOAA, my incident number is: INC0487108 Maybe if we cross reference each other on this, we might be able escalate this a bit. I'm not sure that will happen, but it might be worth a shot. This sure sounds like the same symptoms. My problems started on Saturday AM. How about the rest of you?

**Reply by hanksims** [reply]
@digamesystems Where did you file this ticket? And yes, it was just about exactly that time.

**Reply by digamesystems** [reply]
I started with an email to: nco.ops@noaa.gov. I referenced a link to this GitHub Discussion page and gave a description of the issue. https://github.com/weather-gov/api/discussions/763#discussioncomment-10538461 They got back pretty quickly and then created a ticket. The support link can be found on their API web page: https://www.weather.gov/documentation/services-web-api

**Reply by hanksims** [reply]
Great. I have done that as well, and referenced your incident number.

**Reply by digamesystems** [reply]
@hanksims -- I just got an update to my incident ticket showing your comments. Nice!

**Reply by jah1007** [reply]
We have opened a case with Akamai. From what they have found so far is that there was a change that occurred on 8/31 with how the Linode IPs are advertised. Does that line up for everyone as when this all went sideways?

**Reply by digamesystems** [reply]
It does for me. My service started reporting errors Saturday (the 31st) in the AM. PST

**Reply by digamesystems** [reply]
@jah1007 My ticket with Linode is:
#23587012: Issue With IP Address Allocation

Feel free to share that with them so they can cross reference the issue if it looks useful.

**Reply by berthoud** [reply]
Other Linode user here: Since end of August we have the same problem on requests to api.weather.gov from our Linode servers. Thank you for clarification on this problem. For us, access does work from other computers. The suggestion below, using IPV4 works from our Linode servers.

**Comment by coding-opossum** [comment]
I've been trying to figure out how to connect to/use the NWS API today for my Dakboard (the ones that are built in with the program/service are inaccurate for my location) and keep running into 

> Access Denied
>You don't have permission to access "http://api.weather.gov/points/{lat},{long}" on this server.
>Reference #18.91d7c617.1725413272.637906e
>https://errors.edgesuite.net/18.91d7c617.1725413272.637906e

Is it related? It looks like it but since the reference is different after the 18. I'm not sure. I hope though because otherwise I have no idea what's wrong

**Reply by jah1007** [reply]
Just to be clear, you are providing a lat/lon in that call, correct?

**Reply by coding-opossum** [reply]
Yes, I am. I just didn't want to share my location

**Comment by cdzombak** [comment]
I started seeing the same issue a few days ago, from my home Internet connection:

```
<HTML><HEAD>
<TITLE>Access Denied</TITLE>
</HEAD><BODY>
<H1>Access Denied</H1>

You don't have permission to access "http&#58;&#47;&#47;api&#46;weather&#46;gov&#47;points&#47;42&#46;37&#44;&#45;84&#46;01" on this server.<P>
Reference&#32;&#35;18&#46;50623417&#46;1725373741&#46;49faeeb
<P>https&#58;&#47;&#47;errors&#46;edgesuite&#46;net&#47;18&#46;50623417&#46;1725373741&#46;49faeeb</P>
</BODY>
</HTML>
```

**Reply by digamesystems** [reply]
Interesting. So far in this discussion folks have been seeing the issue when using hosting from Akamai/Linode. Are you trying this from code hosted on your home system?

**Comment by digamesystems** [comment]
Linode is digging into this on their end. Update:

<img width="673" alt="image" src="https://github.com/user-attachments/assets/ed532332-5c18-4104-8d29-5476c75dad66">

**Reply by digamesystems** [reply]
Any progress for other folks?

**Reply by digamesystems** [reply]
A question from our friends at Linode:

<img width="804" alt="image" src="https://github.com/user-attachments/assets/5dee19d7-e90d-4837-8f69-1937eba8f329">

I've tried adding in the header in a call using curl. No luck. Has anyone else made calls using the recommended header with the User-Agent fields?

**Reply by VioletPixel** [reply]
I've had a `User-Agent` header with my email address in place for years. I also tried variations when this started happening and it made no difference.

**Reply by digamesystems** [reply]
@VioletPixel thanks! If you have a ticket in with Linode, feel free to reference mine so we can get some more attention.ticket number : 23587012

**Reply by hanksims** [reply]
Well, I hope I'm not blowing it for myself and everyone else, but...

Setting a User-Agent does not work. But forcing IPv4 does work.

So with cURL, for example:

`curl https://api.weather.gov`

... results in ...

```
<HTML><HEAD>
<TITLE>Access Denied</TITLE>
</HEAD><BODY>
<H1>Access Denied</H1>
 
You don't have permission to access "http&#58;&#47;&#47;api&#46;weather&#46;gov&#47;" on this server.<P>
Reference&#32;&#35;18&#46;570c3417&#46;1725488913&#46;4e6c404
<P>https&#58;&#47;&#47;errors&#46;edgesuite&#46;net&#47;18&#46;570c3417&#46;1725488913&#46;4e6c404</P>
</BODY>
</HTML>
```

... but ...

`curl -4 https://api.weather.gov`

... results in ...

```
{
    "status": "OK"
}
```

**Reply by jaredwsmith** [reply]
I can confirm that forcing IPv4 worked for me as well.

**Reply by VioletPixel** [reply]
Oh, wow, thank you!  Forcing IPv4 works for me as well, both with curl and in my app.  I tried forcing IPv6 earlier, but didn't think it would be using that by default.  Wild!

**Reply by digamesystems** [reply]
Can confirm. So, Linode... are you advertising us as IP v 6?

**Reply by digamesystems** [reply]
@VioletPixel, what language are you using for your app? I'm using python.

**Reply by VioletPixel** [reply]
I'm using PHP.

**Comment by joepperkins** [comment]
The below works for me on windows, even without the used-agent

curl -X GET "https://api.weather.gov/gridpoints/MTR/99,82/forecast" -H "accept: application/geo+json" -H "Feature-Flags: "

I sent the same request to a test server and the curl is adding the header User-Agent: curl/8.8.0 on its own

curl -X GET https://joeperkins.com/test/mgi/index.php -H "accept: application/geo+json" -H "Feature-Flags: "

Host: joeperkins.com
Connection: close
User-Agent: curl/8.8.0
Accept: application/geo+jso

**Comment by digamesystems** [comment]
Ok. We might be getting closer. Thanks to everyone for the help on this!
![image](https://github.com/user-attachments/assets/f80c22ea-c2f6-4efb-a401-790ad7f34cf5)

**Comment by akleinsteiber** [comment]
I'm seeing the same issue from an IP at DigitalOcean that's been working for months with minimal traffic. Started around the same time & forcing IPv4 also works.

Makes me think some configuration has changed on NWS's side that's blocking legitimate traffic, since we have people in the thread having the same issue on Akamai/Linode, home internet, and DigitalOcean.

**Reply by jah1007** [reply]
From what little we have gotten in our ticket, the problem appears to be in how Akamai is classifying IPs. We haven’t made any changes either to our policy or to the origin. I am glad everyone has found a workaround.

**Reply by digamesystems** [reply]
@akleinsteiber, I agree multiple hosting services having this problem needs to be run to ground. Why is a work around suddenly needed? While this fix can be made to work for curl, we are using https requests in python that will need testing. What is to say this work around won't work next week?

**Reply by digamesystems** [reply]
NOAA ops update and my response. 
![image](https://github.com/user-attachments/assets/f1d23450-c585-4f68-8d1a-78f9c24830a5)

**Reply by digamesystems** [reply]
Oops, is ag@gliffen.com associated with NOAA or another user like the rest of us? 
My bad, if so.

**Comment by digamesystems** [comment]
In Python, I'm making https:// GET requests using the requests library. You can restrict it to just using IPV4 with this statement: 

```
import requests
...
requests.packages.urllib3.util.connection.HAS_IPV6 = False
```
My app is working now, but I'd have a much warmer feeling if I had an understanding of why things broke in the first place.. :/ 

Thanks to all for the excellent help! Very much obliged. 

John

**Comment by coding-opossum** [comment]
Hmm, mine still doesn't work. But then again, I'm not doing the coding directly but using the External data/JSON block on the Dakboard OS/site. I've got it set up to use a custom header that's my email (I replaced with generic in the screenshot below) but I'm still getting the error. 
![image](https://github.com/user-attachments/assets/b316f4a8-ebd0-4680-b277-86408a195a2f)

Anyone have any advice? I've done coding but never used dakboard before nor have I used APIs before so I'm also learning on the fly, which makes debugging / problem solving difficult.

**Reply by dwhitemv25** [reply]
I set up a DAKBoard account and pointed that query to my own server to see what it does. It does seem to override the User-Agent in the custom query widget.

But, it looks like DAKBoard is submitting the query from DigitalOcean (over IPv6 to my server). DigitalOcean is a cesspool of abusive activity, and Akamai may not like that. The default user-agent also claims to be IE 5 which is clearly fictitious.

**Reply by coding-opossum** [reply]
I will admit that I have no idea what that means in terms of what I'm trying to do. My coding experience is mostly using python for research and data work. I have no experience with constructing custom monitors (like Dakboard), using APIs, or trying to build widgets / use external data -- I've just been googling my way and using trial and error, haha.

So, is this just not going to work? I've been trying to make/code a weather widget to use on my Dakboard since the ones they link aren't accurate for my location and the NWS API seemed like the best (and most/only personally affordable) way to do so

**Reply by dwhitemv25** [reply]
Technically speaking, what you're doing should work and not generate the 403 error. But as noted in this thread, it seems the API's caching provider has ratcheted up their protective measures and it doesn't think the request from DAKBoard is legitimate and blocks it. At current there's likely little you can do about it until those measures are adjusted or whatever algorithm is in use settles down and removes the block.

That said, if you want the current temperature, that info isn't available on the `/points` endpoint. Best you can do with the API is get weather observations from a nearby airport or other federally-operated weather station. The closest station to the lat/lon provided is a RAWS at 3300 ft elevation, not sure if that is close to you or not (I'm guessing you're in the Hood River Valley from that lat/lon). If you perform the API query in a web browser, ideally with a few more decimal digits (up to 4), you can follow the `observationStations` property link and get the list of available stations in distance order from the point. With the station name, issue the query `/stations/stationIdentifier/observations/latest` to get the latest  observation from that station, with `stationIdentifier` replaced with the value of that property.

For example, for the lan/lon you were using, RXFO3 is the nearest station, and `/stations/RXFO3/observations/latest` is the query to get the most recent conditions from there. There's a `temperature` property in that query that is likely the data you are looking for.

**Reply by VioletPixel** [reply]
Until this issue is fixed you may want/need to switch to a different weather API entirely, like http://pirateweather.net or similar.  I don't think there's anything you can do since you're using a hosted service that doesn't support forcing IPv4, aside from maybe filing a support ticket with the DAKboard folks and asking them if it's something they can do on their end for you.

**Comment by digamesystems** [comment]
I'm not a network security expert. If api.weather.gov is blocking certain domains, why would forcing our requests into IPV4 format "solve" the issue? I'm not complaining that there's a work around, but I want to understand it better. Thanks!

**Reply by VioletPixel** [reply]
It's likely they're blocking a range of IPv6 addresses. Our servers have both IPv6 and IPv4 addresses, and when they connect to something that supports IPv6 (like api.weather.gov) they use it by default, resulting in the issue. Forcing IPv4 makes our servers use a different IP they're not blocking.

**Reply by hanksims** [reply]
I'm not a network security expert either, but I think the idea is that weather.gov -- or, probably, whoever weather.gov contracts with for this service -- has a list of network addresses that it deems to be "bad actors." So it blocks traffic from requests originating from any network address on that list.

For whatever reason, the IPv6 address or addresses we're using to connect to the service from the Linode datacenter is/are on that list. The IPv4 address or addresses is/are not.

**Reply by dwhitemv25** [reply]
I've been affected by a blocklist provider that decided to issue blocks against the containing /48, assuming that the ISP is issuing prefixes in accordance with standards such as [RIPE-690](https://www.ripe.net/publications/docs/ripe-690/). In that case it was a  legacy /64 prefix out of a shared /48; since then that provider has issued /48s to everyone. 

IMO that sort of assumption is unreasonable for a blocklist. It should probably stick to /64s if blocking individual addresses is infeasible. Also, residential and small business ISPs are electing to use /56 delegations as most home users don't need an entire /48. 

This is tricky with cloud providers, which may not be issuing standard-sized prefixes to everyone (or at all), and subnet blocking affects innocent bystanders, which is likely what is being experienced here.

**Reply by digamesystems** [reply]
Thanks, @VioletPixel. That makes a certain sense. What's to stop them from blocking IPv4 addresses from the same domains next week?

This does not provide great comfort to a guy whose users were down for nearly a week. We were lucky this time that it is not the rainy season in California... If inspections are not conducted when they should be during a rain event those users can be subject to substantial fines from state. 

Is there a notification process for these kind of changes to api.weather.gov security policies that I'm not aware of? Mailing list? Web site? Thanks again.

**Reply by dwhitemv25** [reply]
This is a policy from the caching provider, Akamai. Its unlikely NWS had any input on it.

**Reply by digamesystems** [reply]
@dwhitemv25, wait... I'm not sure I understand. Are you saying that the NWS is also using Akamai/Linode?

Recently, in my ticket with them they say that they have not made any changes recently. 

<img width="784" alt="image" src="https://github.com/user-attachments/assets/4680d4a9-55c8-4f46-be2b-db0be50acf19">

**Reply by digamesystems** [reply]
[Support](https://cloud.linode.com/support)
/
[Tickets](https://cloud.linode.com/support/tickets?type=open)
/
#23587012: Issue With IP Address Allocation

**Reply by dwhitemv25** [reply]
It's two different business units. Akamai acquired Linode in 2022. And yes, the irony of one Akamai BU blocking another is not lost on me.

**Reply by digamesystems** [reply]
Wow. Just... wow.

**Reply by digamesystems** [reply]
Update from yesterday evening from the NOAA support team: 
![image](https://github.com/user-attachments/assets/aef1d713-79f5-44a5-8195-c637be78cad3)

Looks like @dwhitemv25's theory might be correct.

**Reply by digamesystems** [reply]
From Linode: 

![image](https://github.com/user-attachments/assets/530e86ba-998f-4922-9a06-370e243b9d97)

**Reply by jah1007** [reply]
The only suggestion provided by Akamai amounts to a game of whack-a-mole on our side unfortunately. This has to be fixed at the Akamai level. We are pushing each day to get an update but so far nothing fruitful as of yet.

**Reply by digamesystems** [reply]
![image](https://github.com/user-attachments/assets/ed380c27-6f56-469d-b6ec-10624176f76b)

**Reply by digamesystems** [reply]
For those whose app might be hard to modify, the IPV6 fix by Linode mentioned above might be worth pursuing.

**Comment by joepperkins** [comment]
IPv6 is a network number and MAC address (simplification) , and modern operating systems today now randomly use a new MAC address every so often to increase privacy. The only option to block a IPv6 address is to bock the enter network that can be many clients depending on network topology.

IPv6 address - Wikipedia<https://en.wikipedia.org/wiki/IPv6_address>
<             Network                 ><        MAC         >
2001:0db8:85a3:0000:0000:8a2e:0370:7334

IPv6 transition mechanism - Wikipedia<https://en.wikipedia.org/wiki/IPv6_transition_mechanism>
Also when going from IPv4 to IPv6 or IPv6 to IPv4 through a common NAT many networks can be involved.

**Comment by digamesystems** [comment]
For those still hanging in on this thread, I received a note from the Linode support team: 

![image](https://github.com/user-attachments/assets/6eb7af8f-9318-4ffc-8d81-17e5d5e44752)

**Reply by jah1007** [reply]
Akamai pushed an update last Thursday which looked to resolve the blocking we were seeing.

**Comment by sullynole** [comment]
Is anyone on this thread still being blocked with 403 errors since 8/31?  If so, please provide the most recent Reference ID from the 403 error so we can work with our CDN to resolve these issues.

**Reply by Ikalike112** [reply]
Reference #18.90f4d517.1729085594.4dcf729a

**Reply by zsloan112** [reply]
Reference Number: #18.8f643017.1729170606.2c47a804

**Reply by sullynole** [reply]
> Reference Number: #18.8f643017.1729170606.2c47a804

Your IP was blocked for exceeding rate limits.  It will auto unblock when rate limits are no longer being exceeded.
